packages/vite/src/node/optimizer/scan.ts

```diff
@@ -264,39 +263,58 @@ function esbuildScanPlugin(
               const src = srcMatch[1] || srcMatch[2] || srcMatch[3]
               js += `import ${JSON.stringify(src)}\n`
             } else if (content.trim()) {
-              // There can be module scripts (`<script context="module">` in Svelte and `<script>` in Vue)
+              // The reason why virtual modules are needed:
+              // 1. There can be module scripts (`<script context="module">` in Svelte and `<script>` in Vue)
               // or local scripts (`<script>` in Svelte and `<script setup>` in Vue)
+              // 2. There can be multiple module scripts in html
               // We need to handle these separately in case variable names are reused between them
+
+              // append imports in TS to prevent esbuild from removing them
+              // since they may be used in the template
+              const contents =
+                content +
+                (loader.startsWith('ts') ? extractImportPaths(content) : '')
+
+              const key = `${path}?id=${scriptId++}`
+
+              if (contents.includes('import.meta.glob')) {
+                scripts[key] = {
+                  // transformGlob already transforms to js
+                  loader: 'js',
+                  contents: await transformGlob(
+                    contents,
+                    path,
+                    config.root,
+                    loader
+                  )
+                }
+              } else {
+                scripts[key] = {
+                  loader,
+                  contents
+                }
+              }
+
+              const virtualModulePath = JSON.stringify(
+                virtualModulePrefix + key
+              )
+
               const contextMatch = openTag.match(contextRE)
               const context =
                 contextMatch &&
                 (contextMatch[1] || contextMatch[2] || contextMatch[3])
-              if (
-                (path.endsWith('.vue') && setupRE.test(openTag)) ||
```

Comment:
shouldn't this vue script setup check still be present?

Reply:
In a Vue file with this PR, all script tags will become `export * from 'virtual-module:...'`. And I believe this should cover Vue's case too since script setup doesn't have exports (?), and there would be no export name clashes.